### PR TITLE
otpilot: Add GPIO handling.

### PR DESCRIPTION
**Pull request by osenft**
This change adds the GPIO driver as well as a GPIO processor that handles
BMC reset monitor events.

### CODE REVIEW

File: userspace/otpilot/src/main.rs [hunk 1]
```diff
@@ -148,9 +154,19 @@ fn run() -> TockResult<()> {
 
     //////////////////////////////////////////////////////////////////////////////
 
+    // We assume that we've already done all boot-time checks at this point.
```
Comment (dann21) [inline userspace/otpilot/src/main.rs:157]: Is there any way to assert this?

File: userspace/otpilot/src/main.rs [hunk 2]
```diff
@@ -148,9 +154,19 @@ fn run() -> TockResult<()> {
...
+
+    // Deassert BMC resets.
+    // TODO(osk): Do something with the result codes.
```
Comment (dann21) [inline userspace/otpilot/src/main.rs:160]: "For now", maybe assert success?

File: userspace/otpilot/src/main.rs [hunk 3]
```diff
@@ -183,9 +199,24 @@ fn run() -> TockResult<()> {
             console_reader::get().allow_read(1)?;
         }
 
+        if gpio::get().have_events() {
+            match gpio_processor.process_gpio_events() {
+                Ok(()) => {}
+                Err(_) => {
+                    // Ignore error from writeln. There's nothing we can do here anyway.
+                    let _ = writeln!(console, "Device: Error processing GPIO event.");
```
Comment (dann21) [inline userspace/otpilot/src/main.rs:207]: Is it possible to add any detail about the event for which processing failed?

File: userspace/otpilot/src/main.rs [hunk 4]
```diff
@@ -183,9 +199,24 @@ fn run() -> TockResult<()> {
...
+        }
+
         if alarm::get().is_expired() {
-            writeln!(console, "Alarm expired.")?;
-            alarm::get().clear()?;
+            match gpio_processor.alarm_expired() {
+                Ok(()) => {}
+                Err(_) => {
+                    // Ignore error from writeln. There's nothing we can do here anyway.
+                    let _ = writeln!(console, "Alarm: Error processing event.");
```
Comment (dann21) [inline userspace/otpilot/src/main.rs:217]: "Error processing event."?  Is that right?

File: userspace/otpilot/src/gpio.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            IS_INITIALIZED = true;
+        }
+        &GPIO_CTRL
+    }
+}
+
+
+fn add_event(events: &Cell<usize>) {
+    let val = events.get();
+    events.set(val + 1);
```
Comment (dann21) [inline userspace/otpilot/src/gpio.rs]: No worry about wraparound here?

File: userspace/otpilot/src/gpio.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+fn delete_event(events: &Cell<usize>) -> usize {
+    let val = events.get();
+    if val > 0 {
```
Comment (dann21) [inline userspace/otpilot/src/gpio.rs]: Hmm.  Should it be OK to silently do nothing here when `val == 0`?

File: userspace/otpilot/src/gpio.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        events.set(val - 1);
+    }
+    val
+}
+
+fn clear_event(events: &Cell<usize>) -> usize {
+    let val = events.get();
+    if val > 0 {
```
Comment (dann21) [inline userspace/otpilot/src/gpio.rs]: Why conditionalize?  Why not always call `events.set(0)` here?

"Conditionals are the source of all complexity" -- rhh, ca. 1983

**Review by dann21 — APPROVED** [review]
LGTM.
Comments are a bunch of nit-picks.

File: userspace/otpilot/src/main.rs (hunk 1 above)
Comment (osenft) [reply userspace/otpilot/src/main.rs:157]: This is more of a "design it in such a way that ..." type of comment.

File: userspace/otpilot/src/main.rs (hunk 2 above)
Comment (osenft) [reply userspace/otpilot/src/main.rs:160]: We need to implement proper "boot time checks" during which we'd want to handle errors like that better, e.g. by resetting Haven or so. I think of this section as a placeholder to implement something.

File: userspace/otpilot/src/main.rs (hunk 3 above)
Comment (osenft) [reply userspace/otpilot/src/main.rs:207]: I created #227 to generally implement better handling in these places.

File: userspace/otpilot/src/main.rs (hunk 4 above)
Comment (osenft) [reply userspace/otpilot/src/main.rs:217]: Yeah - it's an alarm event. I think that's ok.

File: userspace/otpilot/src/gpio.rs (hunk 5 above)
Comment (osenft) [reply userspace/otpilot/src/gpio.rs]: Good call! Fixed.

File: userspace/otpilot/src/gpio.rs (hunk 6 above)
Comment (osenft) [reply userspace/otpilot/src/gpio.rs]: I added a comment explaining what this method is doing. So yes, that's ok.

File: userspace/otpilot/src/gpio.rs (hunk 7 above)
Comment (osenft) [reply userspace/otpilot/src/gpio.rs]: Fair point. Removed.

**Pull request merged**